egs/sad_rats/s5/local/prepare_data.py

```diff
@@ -0,0 +1,168 @@
+Copyright 2020 ARL  (Author: John Morgan)
+Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)  
+
+Prepare rats_sad data for training a speech activity detection system.
+Input comes from the annotations provided with the corpus. 
+Output is written to an RTTM file.
+"""
+
+import sys
+import os
+import argparse
+import subprocess
+import itertools
+from collections import defaultdict
+
+import numpy as np
+import pandas as pd
+
+class Segment:
+    """
+    Field 5 (SAD segment label) can have one of the following values:
+    S  : speech segment
+    NS : non-speech segment
+    NT : "button-off" segment (not transmitted according to PTT log)
+    RX : "button-off" segment (not transmitted according to RMS scan)
+
+    Filenames:
+    transceiver files:   {iiiii}_{jjjjj}_{lng}_{c}.{type}
+    where:
+    {iiiii} is a 5-digit source audio identifier
+    {jjjjj} is a 5-digit transmission session identifier
+    {lng} is one of:
+        alv: Levantine Arabic
+        eng: American English
+        fas: Farsi (Persian)
+        pus: Pashto
+        urd: Urdu
+    {c}    is one of: A B C D E F G H
+    """
+    def __init__(self, fields):
+        self.partition = fields[0]
+        self.reco_id = fields[1]
+        self.start_time = float(fields[2])
+        self.end_time = float(fields[3])
+        self.dur = self.end_time - self.start_time
+        self.sad_label = fields[4]
+        self.sad_provenance = fields[5]
+        # self.speaker_id = fields[6]
+        # self.sid_provenance = fields[7]
+        # self.language_id = fields[8]
+        self.language_id = fields[6]
+        # self.lid_provenance = fields[9]
+        self.lid_provenance = fields[7]
+        # self.transcript = fields[10]
+        # self.transcript_provenance = fields[11]
+
+        rec_info = self.reco_id.split('_')
+        if len(rec_info) == 3:
+            src_audio_id = rec_info[0]
+            lng = rec_info[1]
+            src = rec_info[2]
+            self.spk_id = src_audio_id
+        elif len(rec_info) == 4:
+            src_audio_id = rec_info[0]
+            transmission_session_id = rec_info[1]
+            lng = rec_info[2]
+            channel = rec_info[3]
+            self.spk_id = src_audio_id
+
+
+def groupby(iterable, keyfunc):
+    """Wrapper around ``itertools.groupby`` which sorts data first."""
+    iterable = sorted(iterable, key=keyfunc)
+    for key, group in itertools.groupby(iterable, keyfunc):
+        yield key, group
+
+def read_annotations(file_path):
+    segments = []
+    with open(file_path, 'r') as f:
+        for line in f.readlines():
+            fields = line.strip().split()
+            segments.append(Segment(fields))
+    return segments
+
+
+def find_audios(wav_path, file_list):
+    # Get all .flac file names from audio directory
+    command = 'find %s -name "*.flac"' % (wav_path)
+    wavs = subprocess.check_output(command, shell=True).decode('utf-8').splitlines()
```

Comment:
Using `subprocess()` with `shell=True` can be malicious. Can you check if using `glob()` can work here in a reasonable time? (If it's much slower, then it's okay to keep using subprocess)